README.md
<img src="http://opendatadpc.maps.arcgis.com/sharing/rest/content/items/5c8ef7516b5b4bb19f61037b4cd69015/data" alt="COVID-19" data-canonical-src="http://opendatadpc.maps.arcgis.com/sharing/rest/content/items/5c8ef7516b5b4bb19f61037b4cd69015/data" width="400" />

[Italiano](README.md) - [English](README_EN.md)<br><br>

# Dati COVID-19 Italia

[![GitHub license](https://img.shields.io/badge/License-Creative%20Commons%20Attribution%204.0%20International-blue)](https://github.com/pcm-dpc/COVID-19/blob/master/LICENSE)
[![GitHub commit](https://img.shields.io/github/last-commit/pcm-dpc/COVID-19)](https://github.com/pcm-dpc/COVID-19/commits/master)
 
<<<<<<< HEAD
=======
## Avviso modifiche 
[https://github.com/pcm-dpc/COVID-19/issues/625](https://github.com/pcm-dpc/COVID-19/issues/625)<br><br>
>>>>>>> 44dde7ca11b83b24840315d553018c817698452f

[Sito del Dipartimento della Protezione Civile - Emergenza Coronavirus: la risposta nazionale](http://www.protezionecivile.it/attivita-rischi/rischio-sanitario/emergenze/coronavirus)


Il 31 gennaio 2020, il Consiglio dei Ministri dichiara lo stato di emergenza, per la durata di sei mesi, in conseguenza del rischio sanitario connesso all'infezione da Coronavirus.

Al Capo del Dipartimento della Protezione Civile, Angelo Borrelli, è affidato il coordinamento degli interventi necessari a fronteggiare l'emergenza sul territorio nazionale.  
  
Le principali azioni coordinate dal Capo del Dipartimento sono volte al soccorso e all'assistenza della popolazione eventualmente interessata dal contagio, al potenziamento dei controlli nelle aree aeroportuali e portuali, in continuità con le misure urgenti già adottate dal Ministero della salute, al rientro in Italia dei cittadini che si trovano nei Paesi a rischio e al rimpatrio dei cittadini stranieri nei Paesi di origine esposti al rischio.

Per informare i cittadini e mettere a disposizione i dati raccolti, utili ai soli fini comunicativi e di informazione, il Dipartimento della Protezione Civile ha elaborato un cruscotto geografico interattivo raggiungibile agli indirizzi  [http://arcg.is/C1unv](http://arcg.is/C1unv) (versione desktop) e [http://arcg.is/081a51](http://arcg.is/081a51) (versione mobile) e mette a disposizione, con licenza CC-BY-4.0, le seguenti informazioni aggiornate quotidianamente alle 18:30 (successivamente la conferenza stampa del Capo Dipartimento):

- Dati Andamento nazionale
- Dati json
- Dati regioni
- Dati province
- Schede riepilogative
- Aree
- Note
- Dati contratti DPC forniture

## Avvisi

[Avvisi sui dati andamento COVID-19 Italia](avvisi.md)<br>

## Struttura del repository
```
COVID-19/
│
├── aree/
│   ├── geojson
│   │   ├── dpc-covid-19-ita-aree-comuni.geojson
│   │   ├── dpc-covid19-ita-aree.geojson
│   ├── shp
│   │   ├── dpc-covid19-ita-aree-comuni.dbf
│   │   ├── dpc-covid19-ita-aree-comuni.prj
│   │   ├── dpc-covid19-ita-aree-comuni.shp
│   │   ├── dpc-covid19-ita-aree-comuni.shx
│   │   ├── dpc-covid19-ita-aree.dbf
│   │   ├── dpc-covid19-ita-aree.prj
│   │   ├── dpc-covid19-ita-aree.shp
│   │   ├── dpc-covid19-ita-aree.shx
├── dati-andamento-nazionale/
│   ├── dpc-covid19-ita-andamento-nazionale-*.csv
│   ├── dpc-covid19-ita-andamento-nazionale-latest.csv
│   ├── dpc-covid19-ita-andamento-nazionale.csv
├── dati-contratti-dpc-forniture/
│   ├── dpc-covid19-dati-contratti-dpc-forniture.csv
│   ├── dpc-covid19-dati-pagamenti-contratti-dpc-forniture.csv
│   ├── dati-json
│   │   ├── dpc-covid19-dati-contratti-dpc-forniture.csv
│   │   ├── dpc-covid19-dati-pagamenti-contratti-dpc-forniture.csv
│   ├── file-atti-negoziali
│   │   ├── dpc-contratto-covid19-*.pdf
├── dati-json/
│   ├── dpc-covid19-ita-andamento-nazionale-latest.json
│   ├── dpc-covid19-ita-andamento-nazionale.json
│   ├── dpc-covid19-ita-note-en.json
│   ├── dpc-covid19-ita-note-it.json
│   ├── dpc-covid19-ita-province-latest.json
│   ├── dpc-covid19-ita-province.json
│   ├── dpc-covid19-ita-regioni-latest.json
│   ├── dpc-covid19-ita-regioni.json
├── dati-province/
│   ├── dpc-covid19-ita-province-*.csv
│   ├── dpc-covid19-ita-province-latest.csv
│   ├── dpc-covid19-ita-province.csv
├── dati-regioni/
│   ├── dpc-covid19-ita-regioni-*.csv
│   ├── dpc-covid19-ita-regioni-latest.csv
│   ├── dpc-covid19-ita-regioni.csv
├── note/
│   ├── dpc-covid19-ita-note-en.csv
│   ├── dpc-covid19-ita-note-it.csv
├── schede-riepilogative/
│   ├── province
│   │   ├── dpc-covid19-ita-scheda-province-*.pdf
│   ├── regioni
│   │   ├── dpc-covid19-ita-scheda-regioni-*.pdf
```

## Aggiornamento dei dati

- Dati andamento COVID-19 Italia: ogni giorno alle 18:00<br>
- Dati contratti DPC COVID-19 di fornitura: continua (ogni volta che vengono effettuate operazioni sui contratti)

## Formato dei dati

- [Dati andamento COVID-19 Italia](dati-andamento-covid19-italia.md)<br>
- [Dati contratti DPC COVID-19 di fornitura](dati-contratti-dpc-covid19-fornitura.md)
- [Dati aree misure restrittive COVID19](dati_aree-nazionali-subregionali-misure-restrittive-covid19.md)

## Licenza

[CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/deed.it) - [Visualizza licenza](https://github.com/pcm-dpc/COVID-19/blob/master/LICENSE)


CHANGELOG.md
# Changelog

Tutte le modifiche al progetto, nuove funzionalità e informazioni sono documentate in questo file

## 2020-03-08

- Modifica: "dati-andamento-nazionale" riportando i totali dei dati delle Regioni
- Rimossa: directory "shape-aree-contenimento" e creata "aree" con "shp" e "geojson"

## 2020-03-10

- Aggiornate le Aree secondo il nuovo DPCM 9 Marzo 2020

## 2020-03-11

- Modifica: "Bolzano" e "Trento" rinominati in "P.A. Bolzano" e "P.A. Trento" in "dati-regioni e "dati-province" ("denominazione_regione")
- Modifica: Friuli V. G. rinominata in "dati-regioni e "dati-province" ("denominazione_regione") in "Friuli Venezia Giulia"

## 2020-03-12

- Aggiunto: Codice di Condotta in italiano (CODE_OF_CONDUCT.md) e in inglese (CODE_OF_CONDUCT_EN.md)
- Aggiunto: Changelog in italiano (CHANGELOG.md) e in inglese (CHANGELOG_EN.md)
- Aggiornate le Aree secondo il nuovo DPCM 11 Marzo 2020

## 2020-03-19

- Aggiunti ultimi dati (latest) csv, nelle rispettive directory, per andamento nazionale (dpc-covid19-ita-andamento-nazionale-latest.csv), regioni (dpc-covid19-ita-regioni-latest.csv) e province (dpc-covid19-ita-province-latest.csv)
- Aggiunti ultimi dati (latest) json, nella directory dati-json, per andamento nazionale (dpc-covid19-ita-andamento-nazionale-latest.json), regioni (dpc-covid19-ita-regioni-latest.json) e province (dpc-covid19-ita-province-latest.json)

## 2020-03-25

- Data nel formato ISO8601 UTC
- Aggiunta: "Note" in "dati-regioni", "dati-province" e "dati-andamento-nazionale"
- Aggiunta: dataset "note"

## 2020-03-30

- Modifica: "totale_attualmente_positivi" rinominato in "totale_positivi" (ricoverati_con_sintomi + terapia_intensiva + isolamento domiciliare) in "dati_regioni" e "dati_andamento_nazionale"
- Modifica: "nuovi_attualmente_positivi" rinominato in "variazione_totale_positivi" (totale_attualmente positivi giorno corrente - totale_attualmente positivi giorno precedente) in "dati_regioni" e "dati_andamento_nazionale"
- Aggiunta: "nuovi_positivi" (totale_casi giorno corrente - totale_casi giorno precedente) in "dati_regioni" e "dati_andamento_nazionale"
- Modifica: Regione "Emilia Romagna" rinominato in "Emilia-Romagna" in "dati-regioni" e "dati-province" ("denominazione_regione")

## 2020-04-20

- Aggiunta: "casi_testati", totale dei soggetti sottoposti al test dal 19/04/2020

## 2020-04-21

<<<<<<< HEAD
- Aggiunta: nuovo shape file con misure di contenimento a livello subregionale (province, comuni e frazioni).
=======
- Aggiunta: nuovo shape file con misure di contenimento a livello subregionale (province, comuni e frazioni) con rlativi metadata.
>>>>>>> 44dde7ca11b83b24840315d553018c817698452f

## 2020-04-22

- Aggiunta: nuovo dataset contratti dpc forniture (contratti e pagamenti) in formato csv e json.  [link alla dashboard](https://app.powerbi.com/view?r=eyJrIjoiNzA1YmQ0NDctYzY3ZC00ZTlkLWI1MTQtMThmNTEwNWY3NjM4IiwidCI6IjlhMDZhOTA3LTA2OTUtNDA0YS05NmY4LTRhMWU5YTJmYjQxZCIsImMiOjl9) - [link al dataset ](https://github.com/pcm-dpc/COVID-19/tree/master/dati-contratti-dpc-forniture)

## 2020-04-23

- Aggiunta: metadati del nuovo dataset contratti dpc forniture (contratti e pagamenti) in formato DCAT-AP-IT
- Integrato README con specifiche per il data model dei dataset aree (nazionali e subregionali)

## 2020-04-27

<<<<<<< HEAD
- Aggiunta: aggiornamento al26/4/2020 shape file con misure di contenimento a livello subregionale (province, comuni e frazioni).

## Prossimi aggiornamenti

- API Rest (Json)
- GraphQL

## Ipotesi modifiche
=======
- Aggiunta: aggiornamento al 26/4/2020 shape file e relativi metadata con misure di contenimento a livello subregionale (province, comuni e frazioni).

## 2020-05-04

- Aggiunta: aggiornamento al 4/5/2020 shape file e relativi metadati con misure di contenimento a livello nazionale a seguito del DPCM 26/4/2020.

## 2020-05-05

- API Rest (Json) private beta
- GraphQL private beta

## 2020-05-06

- Aggiunta: aggiornamento al 03/05/2020 shape file e relativi metadata con misure di contenimento a livello subregionale (province, comuni e frazioni).

## Prossimi aggiornamenti
>>>>>>> 44dde7ca11b83b24840315d553018c817698452f

- Cambio codice_regione P.A. Bolzano e P.A. Trento: da 04 a 21 per P.A. Bolzano e da 04 a 22 per P.A. Trento (codice provincia)


util.py
import pandas as pd
import seaborn as sns 
import matplotlib.pyplot as plt
import os
import numpy as np
from scipy.stats import pearsonr
from sklearn.metrics import mean_squared_error
import datetime
import matplotlib.dates as mdates
from matplotlib.dates import DateFormatter
from matplotlib.dates import date2num
import gc
plt.style.use('ggplot')
#suppress annoying warnings about pandas assignment
pd.options.mode.chained_assignment = None
#suppress all warnings
import warnings
warnings.filterwarnings("ignore")
#format x axis with dates
def format_x_date(ax):
    #set ticks every week
    ax.xaxis.set_major_locator(mdates.WeekdayLocator())
    #set major ticks format
    ax.xaxis.set_major_formatter(mdates.DateFormatter('%b %d'))


def get_country_data(selection):
    selection = selection.iloc[:,4:].T
    selection.columns=['Death']
    selection['Death_change']=selection['Death']-selection['Death'].shift(1)
    selection=selection.reset_index()
    selection['index'] = pd.to_datetime(selection['index'])
    selection['Death_change_log']=np.log10(selection['Death_change'])
    selection['Death_change_log']=selection['Death_change_log'].replace(-np.inf,np.nan)
    selection=selection[~selection['Death_change_log'].isna()]
    selection.columns=['date']+list(selection.columns[1:])
    selection['ndate']=date2num(selection['date'])
    return selection

def make_fig1(df):
    plt.style.use('ggplot')
    before = df[df['data']<'2020-03-9']
    before['ndata']=date2num(before['data'])
    after = df[(df['data']>'2020-03-9') & (df['data']<'2020-03-24') ]
    after['ndata']=date2num(after['data'])

    after2 = df[(df['data']>='2020-03-24')  ]
    after2['ndata']=date2num(after2['data'])

    x='ndata'
    y='deceduti_daily_log'
    fig,ax=plt.subplots(figsize=(8,8))
    sns.regplot(x=x, y=y, color="r", data=before,ax=ax,label='before')
    sns.regplot(x=x, y=y, color="b", data=after,ax=ax,label='15 days after ')
    sns.regplot(x=x, y=y, color="black", data=after2,ax=ax,label='25 days after')
    format_x_date(ax)
    xmin, xmax = ax.get_xlim()
    ax.set_xlim(xmin-1,xmax+1)
    #plt.xticks(rotation=70,ha='right')
    plt.xlabel('Date', fontsize='16')
    plt.ylabel('Deaths (log10)',fontsize='16')
    plt.title('Death Trend',fontsize='18')
    plt.legend(title='Social Distancing',fontsize=12)
    plt.savefig('Fig1.png')
    plt.show()
    
    
#plt.style.use('ggplot')

def despine(ax,log_y=False):
    small_add=1.1   
    ymin, ymax = ax.get_ylim()
    if log_y:
        small_add = np.log10(small_add) 
    #print(ymin, ymax)
    #print(small_add)
    #ax.set_ylim(ymin+small_add,ymax)

    ax.spines['left'].set_bounds(ymin+small_add, ymax)
    ax.spines['top'].set_color('none')
    ax.spines['right'].set_color('none')
    xmin, xmax = ax.get_xlim()
    #ax.set_xlim(xmin,xmax)

    ax.spines['bottom'].set_bounds(xmin+0.5, xmax)
    
def lollipop_plot(df, ax, date_col='data', value_col='not_admitted_ratio',
                 color='#007acc',label='Ratio',do_despine=True,
                  do_format_date=True,markersize=5,linewidth=5):
    #make lollipop
    ax.plot(df[date_col].dt.date, df[value_col], "o",
        markersize=markersize, color=color, alpha=0.9,label=label)
    #make stick
    ax.vlines(x=df[date_col].dt.date, ymin=0, 
          ymax=df[value_col], color=color, alpha=0.2, linewidth=linewidth)

    #set ticks every week
    if do_despine:
        despine(ax)
    if do_format_date:
        format_x_date(ax)
    
    return ax

def my_plot(df, ax, date_col='data', 
            #value_3='ricoverati_con_sintomi_change',
            value_2='terapia_intensiva_change',
            value_1='deceduti_daily',
            
            label_1='Deaths',
            label_2="Intensive Care",
            #label_3="Hospital Admissions",
           ):
    #make lollipop
    #ax.plot(df[value_3], df[date_col].dt.date, "o",
    #    markersize=5, color='green', alpha=0.9,label=label_3)    
    ax.plot(df[value_2], df[date_col].dt.date, "o",
        markersize=5, color='blue', alpha=0.9,label=label_2)
    ax.plot(df[value_1], df[date_col].dt.date, "o",
        markersize=5, color='red', alpha=0.9,label=label_1)

    
    xmin=df[[value_1,value_2]].min(axis=1)
    xmax=df[[value_1,value_2]].max(axis=1)
    #make stick
    ax.hlines(y=df[date_col].dt.date, xmin=xmin,
              xmax=xmax, color='black', alpha=0.2, linewidth=2)

    
    #ymin, ymax = ax.get_ylim()
    #ax.set_ylim(ymin,ymax)
    #ax.spines['left'].set_bounds(ymin, ymax-0.5)
    despine(ax)
    #set ticks every 3 days
    #ax.yaxis.set_major_locator(mdates.WeekdayLocator())
    ax.yaxis.set_major_locator(mdates.DayLocator(interval=3))
    #set major ticks format
    ax.yaxis.set_major_formatter(mdates.DateFormatter('%b %d'))

    #ax.spines['top'].set_color('none')
    #ax.spines['right'].set_color('none')
    #xmin, xmax = ax.get_xlim()
    #ax.spines['bottom'].set_bounds(0, xmax-1)
    return ax

def make_fig2(df):
    plt.style.use('default')
    from matplotlib.patches import Ellipse

    fig,axes=plt.subplots(figsize=(16,6
                                  ),ncols=2)
    """
    ax=axes[1]
    ax = my_plot(df, ax)
    df['ndate']=date2num(df['data'])

    ax.set_xlabel('Cases', fontsize='16')
    ax.set_ylabel('Date', fontsize='16')

    ax.annotate('Social Distancing Measures', 
                (df['deceduti_daily'][14]+20, mdates.date2num(df['data'][13])+0.2 ),
                xytext=(50, 2), 
                textcoords='offset points', 
                arrowprops=dict(arrowstyle='-|>'))
    ax.text(-0.1, 1.1, 'B)', horizontalalignment='center', 
                   verticalalignment='center',
                   transform=ax.transAxes,fontsize=16)

    ax.legend(loc=2, bbox_to_anchor=(0.5, 0.3), title='Infected')
    """
    ax=axes[0]
    ax = lollipop_plot(df, ax, date_col='data',
                       value_col='terapia_intensiva',label='Intensive Care')

    ax.legend(loc=2, bbox_to_anchor=(0.1, 0.8), title='Infected')

    ax.set_xlabel('Date', fontsize='16')
    ax.set_ylabel('Admissions', fontsize='16')
    ax.text(-0.1, 1.1, 'A)', horizontalalignment='center', 
                   verticalalignment='center',
                   transform=ax.transAxes,fontsize=16)


    ax=axes[1]
    before = df[df['data']<'2020-03-9']
    before['ndata']=date2num(before['data'])
    after = df[(df['data']>'2020-03-9') & (df['data']<'2020-03-24') ]
    after['ndata']=date2num(after['data'])

    after2 = df[(df['data']>='2020-03-24')  ]
    after2['ndata']=date2num(after2['data'])

    x='ndata'
    y='deceduti_daily_log'
    #fig,ax=plt.subplots(figsize=(8,8))
    sns.regplot(x=x, y=y, color="r", data=before,ax=ax,label='before')
    sns.regplot(x=x, y=y, color="b", data=after,ax=ax,label='15 days after ')
    sns.regplot(x=x, y=y, color="black", data=after2,ax=ax,label='25 days after')


    despine(ax,log_y=True)
    format_x_date(ax)
    xmin, xmax = ax.get_xlim()
    ax.set_xlim(xmin-1,xmax+1)


    ax.set_xlabel('Date', fontsize='16')
    ax.set_ylabel('Deaths (log10)',fontsize='16')
    ax.text(-0.1, 1.1, 'C)', horizontalalignment='center', 
                   verticalalignment='center',
                   transform=ax.transAxes,fontsize=16)

    plt.tight_layout()
    ax.legend(loc=2, bbox_to_anchor=(0.5, 0.3), title='Lockdown Days')
    plt.suptitle('Death vs Intensive Care Admissions',fontsize='18',y=1.1)
    plt.savefig('Fig2.png')
    plt.show()

def make_fig3(df,allDf):
    plt.style.use('default')
    #sns.palplot(sns.color_palette("Accent"))
    palette=sns.color_palette("Accent")
    days_after_lock=15

    fig,ax=plt.subplots(figsize=(8,8))
    temp = df[df['data']>'2020-03-9']
    temp['days_from_lockdown']=np.arange(0,df[df['data']>'2020-03-9'].shape[0])

    sns.regplot(x='days_from_lockdown',y='deceduti_daily_log'
                ,data=temp.iloc[0:days_after_lock],fit_reg=True,color=palette[0],
               marker='.',line_kws={'linestyle':'--'},ci=None) 
    sns.regplot(x='days_from_lockdown',y='deceduti_daily_log',
                data=temp.iloc[days_after_lock:],fit_reg=True,color=palette[0],
                label='Italy',marker='.',ci=None) 


    spain_df = allDf[(allDf['Country/Region']=='Spain')]
    spain_df =  get_country_data(spain_df)
    spain_df = spain_df[spain_df['date']>='2020-03-16']
    spain_df['days_from_lockdown']=np.arange(0,spain_df.shape[0])


    sns.regplot(x='days_from_lockdown',y='Death_change_log'
                ,data=spain_df.iloc[0:days_after_lock],fit_reg=True,color=palette[1],
                marker='.',line_kws={'linestyle':'--'},ci=None) 
    sns.regplot(x='days_from_lockdown',y='Death_change_log',
                data=spain_df.iloc[days_after_lock:],fit_reg=True,color=palette[1],
                label='Spain',marker='.',ci=None) 


    uk_df = allDf[(allDf['Country/Region']=='United Kingdom') & (allDf['Lat']==55.3781)  ]
    uk_df =  get_country_data(uk_df) 
    uk_df = uk_df[uk_df['date']>='2020-03-23']
    uk_df['days_from_lockdown']=np.arange(0,uk_df.shape[0])


    sns.regplot(x='days_from_lockdown',y='Death_change_log'
                ,data=uk_df.iloc[0:days_after_lock],fit_reg=True,color=palette[-1],marker='.',
               line_kws={'linestyle':'--'},ci=None) 
    sns.regplot(x='days_from_lockdown',y='Death_change_log',
                data=uk_df.iloc[days_after_lock:],fit_reg=True,
                color=palette[-1],label='UK',marker='.',ci=None) 





    ax.set_xlim(-2,temp.shape[0]+2)
    ax.set_ylabel('Deaths (log10)',fontsize='16')
    ax.set_xlabel('Days from Lockdown',fontsize='16')
    ax.legend(title='Country ',fontsize=12,loc=2)
    
    despine(ax,log_y=True)
    plt.title('Country Comparison',fontsize='18')
    plt.tight_layout()
    plt.savefig('Fig3.png')

    
def merge_legends(axes):
    lines = []
    labels = []
    for ax in axes:  
        tlines, tlabels = ax.get_legend_handles_labels()
        lines+=tlines
        labels+=tlabels
    return lines,labels
    